### PR TITLE
Remove dependency on Hosting.Abstractions

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
This removes the dependency on `Microsoft.AspNetCore.Hosting.Abstractions`.

### CODE REVIEW

File: src/Microsoft.Extensions.Localization/IApplicationNameProvider.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Microsoft.Extensions.Localization
```
Comment (JunTaoLuo) [inline src/Microsoft.Extensions.Localization/IApplicationNameProvider.cs]: Copyright

File: src/Microsoft.Extensions.Localization/IApplicationNameProvider.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.Extensions.Localization
+{
+    public interface IApplicationNameProvider
```
Comment (rynowak) [inline src/Microsoft.Extensions.Localization/IApplicationNameProvider.cs]: This problem with this approach is that your `AddLocalization` methods no longer work out of the box. This removes the dependency on Hosting, but replaces it with an implicit dependency on something else. 

Why not add the application name to `LocalizationOptions` and do both of:
- define the behavior to something sane when application name isn't set
- teach MVC to set the application name for your when localization is in use

This doesn't require any new abstractions or plumbing, and has the advantage that localization can be used outside of MVC.

Comment (rynowak) [inline src/Microsoft.Extensions.Localization/IApplicationNameProvider.cs]: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Localization/Internal/MvcLocalizationServices.cs

**Comment by hishamco** [comment]
What if we didn't allow null-ability for `location` here https://github.com/aspnet/Localization/blob/07e774146306d7f838aea279655b9a6058e1c7a1/src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs#L158 instead of https://github.com/aspnet/Localization/blob/07e774146306d7f838aea279655b9a6058e1c7a1/src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs#L165

File: src/Microsoft.Extensions.Localization/IApplicationNameProvider.cs (hunk 2 above)
Comment (Eilon) [inline src/Microsoft.Extensions.Localization/IApplicationNameProvider.cs]: Hmm, I do wonder - which of the Entropy (or other?) samples need to be updated due to this? I admit I hadn't quite considered those implications.

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 3]
```diff
@@ -1,9 +1,9 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved. 
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
```
Comment (Eilon) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Needed?

Comment (ryanbrandenburg) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Needed for when we `Assert.Throw<ArgumentNullException>`

Comment (hishamco) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: That's what I wanna say 😄

**Comment by hishamco** [comment]
Seems the final changes looks fine 👍, well done!!

**Comment by hishamco** [comment]
@ryanbrandenburg seems we need to remove these too https://github.com/aspnet/Localization/blob/6f3a68d80dda9681e2fc3ff0020ba38291d41aae/src/Microsoft.Extensions.Localization/baseline.netcore.json#L320-L323, https://github.com/aspnet/Localization/blob/6f3a68d80dda9681e2fc3ff0020ba38291d41aae/src/Microsoft.Extensions.Localization/baseline.net45.json#L320-L323

**Comment by Eilon** [comment]
@hishamco we don't update the baselines; rather, we just add exclusions in a separate list.

**Comment by hishamco** [comment]
Thanks, because I thought that may required for the docs, coz tell now I don't have any idea about these files 😄

File: src/Microsoft.Extensions.Localization/exceptions.net45.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+[
+    {
+        "OldTypeId": "public class Microsoft.Extensions.Localization.ResourceManagerStringLocalizerFactory : Microsoft.Extensions.Localization.IStringLocalizerFactory",
+        "OldMemberId": "public .ctor(Microsoft.AspNetCore.Hosting.IHostingEnvironment hostingEnvironment, Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Localization.LocalizationOptions> localizationOptions)",
+        "NewTypeId": "public class Microsoft.Extensions.Localization.ResourceManagerStringLocalizerFactory : Microsoft.Extensions.Localization.IStringLocalizerFactory",
+        "NewMemberId": "public .ctor(Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Localization.LocalizationOptions> localizationOptions)",
```
Comment (Eilon) [inline src/Microsoft.Extensions.Localization/exceptions.net45.json]: @javiercn - how exactly does the tool know that this ctor was "modified" as opposed to an old one got deleted, and a new one showed up? Is that not fragile?

Comment (javiercn) [inline src/Microsoft.Extensions.Localization/exceptions.net45.json]: We have a modification kind

Comment (javiercn) [inline src/Microsoft.Extensions.Localization/exceptions.net45.json]: You set it to Removal, Modification or Adition

Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.Localization/exceptions.net45.json]: @Eilon is that enough of an explanation and this is approved now?

Comment (Eilon) [inline src/Microsoft.Extensions.Localization/exceptions.net45.json]: Yep.

**Comment by hishamco** [comment]
The AppVeyor failed!!

**Comment by ryanbrandenburg** [comment]
An announcement was made for this breaking change in aspnet/Announcements#218.

**Pull request merged**